### PR TITLE
feat(x/incentives): min value param for epoch distribution

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #7610

## What is the purpose of the change

Completes https://www.mintscan.io/osmosis/proposals/733

Adds a min value to trigger epoch incentive distributions and sets to 0.01 OSMO.

If the underlying token being distributed does not have a path registered in protorev OR the underlying being distributed is less than the new min value parameter, the distribution for that is skipped.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new parameter to set a minimum value for distribution eligibility in the incentives module.
  - Added caching for minimum distribution values to optimize performance.
- **Enhancements**
  - Implemented a new keeper for protocol revisions.
  - Enhanced distribution logic to handle non-base denomination rewards and insufficient rewards scenarios.
- **Tests**
  - Expanded test coverage to include checks for the new minimum value for distribution, non-base denomination rewards handling, and user lock configurations.
- **Documentation**
  - Updated CHANGELOG.md to reflect new parameter addition for epoch distribution minimum value.

### CODE REVIEW

File: proto/osmosis/incentives/params.proto [hunk 1]
```diff
@@ -44,4 +44,11 @@ message Params {
     (gogoproto.stdduration) = true,
     (gogoproto.moretags) = "yaml:\"internal_uptime\""
   ];
+  // min_value_for_distribution is the minimum amount a token must be worth
+  // in order to be eligible for distribution. If the token is worth
+  // less than this amount (or the route between the two denoms is not
+  // registered), it will not be distributed and is forfeited to the remaining
+  // distributees that are eligible.
```
Comment (mattverse) [inline proto/osmosis/incentives/params.proto:51]:
```suggestion
  // distributes that are eligible.
```

File: x/cosmwasmpool/cosmwasm/msg/transmuter/transmuter_test.go [hunk 2]
```diff
@@ -50,6 +50,9 @@ func (s *TransmuterSuite) TestFunctionalTransmuter() {
 		expectedDenomSuffix  = "/transmuter/poolshare"
 	)
 
+	// Set base denom
+	s.App.IncentivesKeeper.SetParam(s.Ctx, incentivetypes.KeyMinValueForDistr, sdk.NewCoin("uosmo", osmomath.NewInt(10000)))
```
Comment (mattverse) [inline x/cosmwasmpool/cosmwasm/msg/transmuter/transmuter_test.go:54]: What's the reason for this change?

**Review by mattverse — COMMENTED** [review]
LGTM, so this is happening in epoch as well right? I'm pretty sure it is the case, but do you happen to have any benchmark or brief reasoning to support that this calculation / logic is less intense computational wise compared to us just distributing all tokens?

File: x/cosmwasmpool/cosmwasm/msg/transmuter/transmuter_test.go (hunk 2 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/cosmwasm/msg/transmuter/transmuter_test.go:54]: The asset being distributed via gauges either needs to be the denom set in the param store OR have a pool linking the denom to the denom set in the param store. So the option here is to either set the param to the denom in the test or create a pool linking the denom in the test to the denom in the param store.

**Review by mattverse — APPROVED** [review]

File: x/incentives/keeper/distribute.go [hunk 3]
```diff
@@ -722,6 +733,54 @@ func (k Keeper) distributeInternal(
...
+						// Check if the value is worth enough in the token to be distributed.
+						if amtInt.LT(minTokenRequiredForDistr.Amount) {
+							// The value is not worth enough, continue
+							continue
+						}
+					} else {
+						// Cache hit, use the value
+
+						// Check if the underlying is worth enough in the token to be distributed.
+						if amtInt.LT(value) || value.IsZero() {
```
Comment (nicolaslara) [inline x/incentives/keeper/distribute.go:778]: We probably want to split these two and add a comment to make clear what the zero value represents.

```suggestion
						// This route does not exist in protorev so a zero value has been added when a cache miss occurred 
						if value.IsZero() {
							continue
						}
						// Check if the underlying is worth enough in the token to be distributed.
						if amtInt.LT(value) {
```

**Review by nicolaslara — APPROVED** [review]
Added a style/docs suggestion, but other than that this LGTM

File: x/incentives/keeper/distribute.go (hunk 3 above)
Comment (czarcas7ic) [reply x/incentives/keeper/distribute.go:778]: Good suggestion

**Pull request merged**